File: frontend/src/api/equipment.ts
// api/equipment.ts — 设备运行参数 API（DB-002 对应端点）
import client from './client'

// ── 响应类型 ──────────────────────────────────────────────────────────────────

export interface CalciumReactorState {
  id?: string
  tank_id?: string
  flow_rate: number | null
  target_ph: number | null
  outlet_kh: number | null
  updated_at?: string
}

export interface DosingPumpChannel {
  id?: string
  tank_id?: string
  channel_name: string
  dose_g_per_time: number   // g/次
  times_per_day: number     // 次/day
  daily_dose_g: number      // g/day = dose_g_per_time × times_per_day
  updated_at?: string
}

export interface EquipmentResponse {
  calcium_reactor: CalciumReactorState | null
  dosing_channels: DosingPumpChannel[]
}

export interface EquipmentTuningLog {
  id: string
  tank_id: string
  /** equipment = 本次调参汇总（新格式），calcium_reactor/dosing_pump = 旧格式兼容 */
  device_type: 'calcium_reactor' | 'dosing_pump' | 'equipment'
  param_name: string
  old_value: string | null
  new_value: string
  changed_at: string
}

export interface TuningLogsResponse {
  logs: EquipmentTuningLog[]
}

// ── 请求类型 ──────────────────────────────────────────────────────────────────

export interface UpdateEquipmentInput {
  calcium_reactor?: {
    flow_rate?: number | null
    target_ph?: number | null
    outlet_kh?: number | null
  }
  dosing_channels?: {
    channel_name: string
    dose_g_per_time: number
    times_per_day: number
  }[]
}

// ── API 方法 ──────────────────────────────────────────────────────────────────

export const equipmentApi = {
  /** 获取鱼缸设备当前运行参数（钙反 + 所有滴定通道） */
  getEquipment: (tankId: string) =>
    client.get<EquipmentResponse>(`/tanks/${tankId}/equipment`).then(r => r.data),

  /** 调参：UPSERT 钙反/滴定状态，自动写入调参日志 */
  updateEquipment: (tankId: string, input: UpdateEquipmentInput) =>
    client.put<EquipmentResponse>(`/tanks/${tankId}/equipment`, input).then(r => r.data),

  /** 获取调参历史日志，按 changed_at 倒序 */
  getTuningLogs: (tankId: string, limit = 200) =>
    client.get<TuningLogsResponse>(`/tanks/${tankId}/equipment/tuning-logs`, {
      params: { limit },
    }).then(r => r.data),

  /** 删除指定调参日志 */
  deleteTuningLog: (tankId: string, logId: string) =>
    client.delete(`/tanks/${tankId}/equipment/tuning-logs/${logId}`),
}

File: backend/internal/handler/equipment.go
// handler/equipment.go — 设备运行参数 HTTP 处理层
//
// 路由（全部需要 JWT）：
//   GET    /api/tanks/:id/equipment                        — 读取钙反 + 滴定通道当前状态
//   PUT    /api/tanks/:id/equipment                        — 调参（UPSERT 状态 + 自动写日志）
//   GET    /api/tanks/:id/equipment/tuning-logs            — 调参历史日志（倒序）
//   DELETE /api/tanks/:id/equipment/tuning-logs/:logId     — 删除指定调参日志
package handler

import (
	"net/http"
	"strconv"

	"github.com/fuqis/reefmatrix/internal/middleware"
	"github.com/fuqis/reefmatrix/internal/repository"
	"github.com/fuqis/reefmatrix/internal/service"
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

type EquipmentHandler struct {
	svc   *service.EquipmentService
	authz repository.TankAuthz
}

func NewEquipmentHandler(svc *service.EquipmentService, authz repository.TankAuthz) *EquipmentHandler {
	return &EquipmentHandler{svc: svc, authz: authz}
}

// tankIDAndAuth 解析 :id 路径参数并校验当前用户是否为鱼缸 owner
// 成功时返回 tankID；失败时已写 JSON 响应，调用方直接 return
func (h *EquipmentHandler) tankIDAndAuth(c *gin.Context) (uuid.UUID, bool) {
	userID := middleware.GetUserID(c)
	tankID, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid tank id"})
		return uuid.Nil, false
	}
	if !h.authz.OwnsTank(userID, tankID) {
		c.JSON(http.StatusForbidden, gin.H{"error": "无权访问该鱼缸"})
		return uuid.Nil, false
	}
	return tankID, true
}

// ── GET /api/tanks/:id/equipment ─────────────────────────────────────────────

func (h *EquipmentHandler) GetEquipment(c *gin.Context) {
	tankID, ok := h.tankIDAndAuth(c)
	if !ok {
		return
	}
	resp, err := h.svc.GetEquipment(tankID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, resp)
}

// ── PUT /api/tanks/:id/equipment ─────────────────────────────────────────────
//
// Body：
//   {
//     "calcium_reactor": { "flow_rate": 35, "target_ph": 6.6, "outlet_kh": 48 },
//     "dosing_channels": [
//       { "channel_name": "Ca补充", "daily_dose_g": 2.5 }
//     ]
//   }
//
// calcium_reactor / dosing_channels 均可选，只传其中一个也合法。

func (h *EquipmentHandler) UpdateEquipment(c *gin.Context) {
	tankID, ok := h.tankIDAndAuth(c)
	if !ok {
		return
	}

	// dosing_channels 用 *json.RawMessage 先检测是否存在，再解析内容
	// 以区分「未传」(nil = 不操作) vs「传了空数组」([]= 删除全部通道)
	var raw struct {
		CalciumReactor *service.CalciumReactorInput  `json:"calcium_reactor"`
		DosingChannels *[]service.DosingChannelInput  `json:"dosing_channels"`
	}
	if err := c.ShouldBindJSON(&raw); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// 统一调参：一次 PUT 只写 ONE 条调参日志；dosing 为 nil 时不触发通道同步
	if err := h.svc.UpdateEquipmentBatch(tankID, raw.CalciumReactor, raw.DosingChannels); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	// 返回更新后的完整状态
	resp, err := h.svc.GetEquipment(tankID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, resp)
}

// ── GET /api/tanks/:id/equipment/tuning-logs ─────────────────────────────────
//
// Query params：
//   limit  最多返回条数（默认 50，最大 200）

func (h *EquipmentHandler) GetTuningLogs(c *gin.Context) {
	tankID, ok := h.tankIDAndAuth(c)
	if !ok {
		return
	}

	limit := 50
	if l := c.Query("limit"); l != "" {
		if v, err := strconv.Atoi(l); err == nil && v > 0 && v <= 200 {
			limit = v
		}
	}

	logs, err := h.svc.GetTuningLogs(tankID, limit)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"logs": logs})
}

// ── DELETE /api/tanks/:id/equipment/tuning-logs/:logId ───────────────────────

func (h *EquipmentHandler) DeleteTuningLog(c *gin.Context) {
	tankID, ok := h.tankIDAndAuth(c)
	if !ok {
		return
	}
	logID, err := uuid.Parse(c.Param("logId"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid log id"})
		return
	}
	if err := h.svc.DeleteTuningLog(tankID, logID); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "记录不存在或无权删除"})
		return
	}
	c.Status(http.StatusNoContent)
}
